modules/javafx.graphics/src/main/java/com/sun/javafx/util/DataURI.java

```diff
@@ -208,4 +203,83 @@ public int hashCode() {
         return result;
     }
 
+    /**
+     * Decodes percent-encoded text as specified by RFC 3986, section 2.1
+     * This method does not make any assumptions about the allowed character set.
+     *
+     * @param input the input string
+     * @return the decoded byte array
+     */
+    private static byte[] decodePercentEncoding(String input) {
+        enum ExpectedCharacter {
+            DEFAULT,
+            FIRST_HEX_DIGIT,
+            SECOND_HEX_DIGIT
+        }
+
+        ExpectedCharacter expectedCharacter = ExpectedCharacter.DEFAULT;
+        byte[] output = new byte[computePayloadSize(input)];
```

Comment:
just a thought: what if the url contains no %?  in this case one can simply return the input string.
I think we could return -1 from computePayloadSize() to indicate that condition.

Reply:
The return value is not a string, it's a byte array. So we need to do the conversion anyway, which is what this method is doing quite efficiently by now. It's probably not useful to use another conversion method instead, at least not in terms of performance.